File: ui/damage-calculator.test.js
/**
 * Zero-dependency Node.js test runner for damage-calculator.js.
 *
 * Reads the same test-cases JSON used by the Java DamageCalculatorTest and
 * verifies that the JS port produces identical results (within ±0.001).
 *
 * Usage:  node --experimental-vm-modules ui/damage-calculator.test.js
 *    or:  node ui/damage-calculator.test.js          (Node 22+)
 */

import { calculate } from './damage-calculator.js';
import { readFileSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const casesPath = resolve(__dirname, 'test-cases.json');
const cases = JSON.parse(readFileSync(casesPath, 'utf-8'));

const TOLERANCE = 0.001;
let passed = 0;
let failed = 0;

function approxEqual(a, b, label) {
    if (Math.abs(a - b) > TOLERANCE) {
        return `  FAIL ${label}: expected ${b}, got ${a} (diff ${Math.abs(a - b).toFixed(6)})`;
    }
    return null;
}

for (const testCase of cases) {
    const errors = [];

    // Build inputs in the same shape the UI sends
    const inputs = {
        rawDamage:          testCase.mob.rawDamage,
        starLevel:          testCase.mob.starLevel,
        extraDamagePercent: testCase.mob.extraDamagePercent ?? 0,
        difficulty:         testCase.difficulty,
        maxHealth:          testCase.player.maxHealth,
        blockingSkill:      testCase.player.blockingSkill,
        blockArmor:         testCase.player.blockArmor,
        armor:              testCase.player.armor,
    };

    // Resolve parry multiplier — prefer explicit, fall back to parryBonus enum
    if (testCase.player.parryMultiplier != null) {
        inputs.parryMultiplier = testCase.player.parryMultiplier;
    } else {
        inputs.parryBonus = testCase.player.parryBonus;
    }

    let data;
    try {
        data = calculate(inputs);
    } catch (error) {
        errors.push(`  EXCEPTION: ${error.message}`);
        console.log(`✗ ${testCase.name}`);
        errors.forEach(errorLine => console.log(errorLine));
        failed++;
        continue;
    }

    // Pick the scenario that the test case targets
    const result = !testCase.useShield ? data.noShield : testCase.isParry ? data.parry : data.block;
    const expected = testCase.expected;

    // Assertions
    const checks = [
        approxEqual(data.baseRawDamage,             expected.baseRawDamage,           'baseRawDamage'),
        approxEqual(data.effectiveRawDamage,         expected.effectiveRawDamage,      'effectiveRawDamage'),
        approxEqual(result.blockReducedDamage,    expected.blockReducedDamage,   'blockReducedDamage'),
        approxEqual(result.finalReducedDamage,       expected.finalReducedDamage,      'finalReducedDamage'),
        approxEqual(result.remainingHealth,           expected.remainingHealth,         'remainingHealth'),
    ];

    // Stagger (string comparison)
    if (result.stagger !== expected.stagger) {
        checks.push(`  FAIL stagger: expected ${expected.stagger}, got ${result.stagger}`);
    }
    // minHealthForNoStagger (exact int)
    if (result.minHealthForNoStagger !== expected.minHealthForNoStagger) {
        checks.push(`  FAIL minHealthForNoStagger: expected ${expected.minHealthForNoStagger}, got ${result.minHealthForNoStagger}`);
    }

    const failures = checks.filter(Boolean);
    if (failures.length) {
        console.log(`✗ ${testCase.name}`);
        failures.forEach(failure => console.log(failure));
        failed++;
    } else {
        console.log(`✓ ${testCase.name}`);
        passed++;
    }
}

console.log(`\n${passed} passed, ${failed} failed, ${cases.length} total`);
process.exit(failed > 0 ? 1 : 0);

File: src/test/java/valheim/calculator/DamageCalculatorTest.java
package valheim.calculator;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import valheim.calculator.console.ResultPrinter;
import valheim.calculator.core.DamageCalculator;
import valheim.calculator.core.DamageResult;
import valheim.calculator.core.GameDifficulty;
import valheim.calculator.core.MobStats;
import valheim.calculator.core.ParryBonus;
import valheim.calculator.core.PlayerStats;
import valheim.calculator.core.StaggerResult;

import java.io.InputStream;
import java.util.Arrays;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class DamageCalculatorTest {

    @ParameterizedTest(name = "{0}")
    @MethodSource("loadTestCases")
    void calculate(TestCase tc) {
        MobStats mob = MobStats.builder()
                .rawDamage(tc.mob.rawDamage)
                .starLevel(tc.mob.starLevel)
                .extraDamagePercent(tc.mob.extraDamagePercent != null ? tc.mob.extraDamagePercent : 0.0)
                .build();

        PlayerStats player = PlayerStats.builder()
                .maxHealth(tc.player.maxHealth)
                .blockingSkill(tc.player.blockingSkill)
                .blockingArmor(tc.player.blockingArmor)
                .armor(tc.player.armor)
                .parryMultiplier(resolveParryMultiplier(tc.player))
                .build();

        GameDifficulty difficulty = GameDifficulty.valueOf(tc.difficulty);

        DamageResult noShield = DamageCalculator.calculate(player, mob, difficulty, false, false);
        DamageResult block    = DamageCalculator.calculate(player, mob, difficulty, true,  false);
        DamageResult parry    = DamageCalculator.calculate(player, mob, difficulty, true,  true);

        ResultPrinter.printTable(Arrays.asList(noShield, block, parry), difficulty, tc.mob.starLevel,
                mob.rawDamage(), mob.getEffectiveRawDamage(difficulty));

        DamageResult result = !tc.useShield ? noShield : tc.isParry ? parry : block;

        ExpectedOutput exp = tc.expected;
        assertAll(
                () -> assertEquals(exp.baseRawDamage,           mob.rawDamage(),                        0.001, "baseRawDamage"),
                () -> assertEquals(exp.effectiveRawDamage,      mob.getEffectiveRawDamage(difficulty),  0.001, "effectiveRawDamage"),
                () -> assertEquals(exp.blockingReducedDamage,   result.blockingReducedDamage(),         0.001, "blockingReducedDamage"),
                () -> assertEquals(exp.finalReducedDamage,      result.finalReducedDamage(),            0.001, "finalReducedDamage"),
                () -> assertEquals(exp.remainingHealth,         result.remainingHealth(),               0.001, "remainingHealth"),
                () -> assertEquals(StaggerResult.valueOf(exp.stagger), result.stagger(),                       "stagger"),
                () -> assertEquals(exp.minHealthForNoStagger,   result.minHealthForNoStagger(),                "minHealthForNoStagger")
        );
    }

    static Stream<Arguments> loadTestCases() throws Exception {
        ObjectMapper mapper = new ObjectMapper();
        try (InputStream is = DamageCalculatorTest.class.getResourceAsStream("/damage-calculator-test-cases.json")) {
            return Arrays.stream(mapper.readValue(is, TestCase[].class)).map(Arguments::of);
        }
    }

    // ── Inner data classes ─────────────────────────────────────────────────

    static class TestCase {
        public String name;
        public MobInput mob;
        public PlayerInput player;
        public String difficulty;
        public boolean useShield;
        public boolean isParry;
        public ExpectedOutput expected;

        @Override public String toString() { return name; }
    }

    static class MobInput {
        public double rawDamage;
        public int starLevel;
        public Double extraDamagePercent;
    }

    static class PlayerInput {
        public double maxHealth;
        public double blockingSkill;
        public double blockingArmor;
        public double armor;
        public String parryBonus;
        public Double parryMultiplier;
    }

    private static double resolveParryMultiplier(PlayerInput player) {
        if (player.parryMultiplier != null) {
            return player.parryMultiplier;
        }
        return ParryBonus.valueOf(player.parryBonus).multiplier();
    }

    static class ExpectedOutput {
        public double baseRawDamage;
        public double effectiveRawDamage;
        public double blockingReducedDamage;
        public double finalReducedDamage;
        public double remainingHealth;
        public String stagger;
        public int minHealthForNoStagger;
    }
}
